### PR TITLE
Add pre-commit plus a GItHub Action to run it

### PR DESCRIPTION
**Pull request by cclauss**
As discussed at https://github.com/eclipse/paho.mqtt.python/pull/718#discussion_r1432593177

@akx Your review, please.

Using https://pre-commit.ci is even better than using GitHub Actions for pre-commit.

% `pre-commit run --all-files`
```
[INFO] Initializing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Initializing environment for https://github.com/charliermarsh/ruff-pre-commit.
[INFO] Initializing environment for https://github.com/abravalheri/validate-pyproject.
[INFO] Initializing environment for https://github.com/abravalheri/validate-pyproject:.[all].
[INFO] Installing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/charliermarsh/ruff-pre-commit.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/abravalheri/validate-pyproject.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
check builtin type constructor use.......................................Passed
check toml...............................................................Passed
check xml............................................(no files to check)Skipped
check yaml...............................................................Passed
ruff.....................................................................Passed
Validate pyproject.toml..................................................Passed
```

### CODE REVIEW

File: .github/workflows/precommit.yml
Comment (akx) [inline .github/workflows/precommit.yml:21]:
```suggestion
      - uses: actions/setup-python@v5
        with:
          python-version: 3
      - uses: pre-commit/action@v3.0.0
```

File: .github/workflows/precommit.yml
Comment (akx) [inline .github/workflows/precommit.yml:20]: Installing the Git hook is unnecessary in CI :)

**Review by akx — COMMENTED** [review]
This really boils down to maintainer (@PierreF) preference. #770 already set up running Ruff with `tox`, which is also fine. Left a comment about https://github.com/pre-commit/action in case you weren't aware of it.

File: .github/workflows/precommit.yml
Comment (cclauss) [reply .github/workflows/precommit.yml:21]: https://github.com/pre-commit/action is in maintenance-only mode and will not be accepting new features so I tend not to use it.

File: .github/workflows/precommit.yml
Comment (cclauss) [reply .github/workflows/precommit.yml:20]:
```suggestion
```

File: .github/workflows/precommit.yml
Comment (akx) [reply .github/workflows/precommit.yml]: I know it's in maintenance-only mode, but it works – it's also just, well, [20 lines](
https://github.com/pre-commit/action/blob/main/action.yml).

I don't prefer pre-commit.ci because it's an one-man project (and for projects where "auto-fix PRs" is enabled, they tend to mess with my workflow).

Comment (cclauss) [reply .github/workflows/precommit.yml]: If auto-fix PRs are messing up your workflow then you do not have pre-commit properly installed locally.

File: .github/workflows/precommit.yml
Comment (PierreF) [inline .github/workflows/precommit.yml:18]: To my understanding we could use
```suggestion
      - use: pre-commit/action@v3.0.0
```

Which mostly do the same (pip install and pre-commit run) but also add a cache.

Even if only in maintenance, if the version 3.0.0 had an issue, we could just copy their actions.yaml and fix it here.

**Review by PierreF — APPROVED** [review]
This add another tools for checking code, but I think it's okay:
* tox -e py (or just tox / tox -e py38): is here to run Python tests. And might took some time
* tox -e lint will be kept for slow linter
* pre-commit will be used for faster linter

Those linters/tests could still run locally and in Github actions. Which fit the needs of everyone:
* the one that want to run it locally (like me) before pushing to Github
* all PRs and master is also automatically checked by Github

File: .github/workflows/precommit.yml
Comment (PierreF) [inline .github/workflows/precommit.yml:20]:
```suggestion
      - uses: pre-commit/action@v3.0.0
```

Oups, I made a typo in my previous suggestion

**Comment by PierreF** [comment]
The error isn't added by the PR but already present on master. I will fix it once merged.

**Review by akx — COMMENTED** [review]
Weak review comment, can be addressed later as well: If this is adopted (and I'm for it, in general), `tox -e lint` should just run `pre-commit run --all-files` to avoid drift between multiple lint styles.

File: .github/workflows/precommit.yml
Comment (akx) [reply .github/workflows/precommit.yml]: @cclauss Well, for a repo like this where every commit should have a `--signoff` to adhere to a CLA, the pre-commit commit wouldn't be appropriate, etc. 😁

**Pull request merged**
